File: native/android/NativeMeshTransport.kt
package com.resqnet.mesh

/**
 * Android production transport boundary.
 *
 * Implement with BluetoothLeAdvertiser/BluetoothLeScanner and a foreground
 * service. Wi-Fi Direct uses WifiP2pManager and a TCP socket carrying the same
 * MeshFrame bytes. The app must persist the outbox and dedupe packet ids.
 */
interface NativeMeshTransport {
    suspend fun start(): TransportState
    suspend fun stop()
    suspend fun send(peerId: String, frame: ByteArray): SendResult
    fun onFrame(listener: (peerId: String, frame: ByteArray) -> Unit): Closeable
}

interface Closeable {
    fun close()
}

data class SendResult(
    val peerId: String,
    val accepted: Boolean,
    val transport: TransportKind,
)

enum class TransportKind { BLE, WIFI_DIRECT }

enum class TransportState { READY, PERMISSION_NEEDED, UNAVAILABLE }

/** Required Android permissions/features for the host application. */
object NativeMeshRequirements {
    val permissions = listOf(
        "android.permission.BLUETOOTH_SCAN",
        "android.permission.BLUETOOTH_CONNECT",
        "android.permission.BLUETOOTH_ADVERTISE",
        "android.permission.ACCESS_FINE_LOCATION",
        "android.permission.NEARBY_WIFI_DEVICES",
        "android.permission.FOREGROUND_SERVICE_CONNECTED_DEVICE",
    )
}

File: native/ios/NativeMeshTransport.swift
import Foundation

/// iOS production transport boundary.
///
/// Implement BLE with CoreBluetooth and local peer networking with
/// MultipeerConnectivity. Both transports carry the UTF-8 MeshFrame bytes
/// defined in shared/src/meshFrame.ts.
public protocol NativeMeshTransport {
    func start() async throws -> TransportState
    func stop()
    func send(peerID: String, frame: Data) async throws -> SendResult
    @discardableResult
    func onFrame(_ listener: @escaping (_ peerID: String, _ frame: Data) -> Void) -> AnyObject
}

public enum TransportKind: String {
    case bluetooth
    case multipeer
}

public enum TransportState {
    case ready
    case permissionNeeded
    case unavailable
}

public struct SendResult {
    public let peerID: String
    public let accepted: Bool
    public let transport: TransportKind
}

public enum NativeMeshRequirements {
    public static let bluetoothServiceUUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
    public static let bluetoothTxUUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
    public static let bluetoothRxUUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"
    public static let multipeerServiceType = "resqnet-mesh"
}
